RISC-V: Support DTBs with one interrupt value per cell

axel-h opened the pull request:
Seems on ARM this was always used with device trees that have 3 values per interrupt or extended interrupts that have just one interrupt. Fix the handling for extended interrupts to allow more than one. Add support for device trees (e.g. RISC-V HiFive) that use only one value per interrupt. This is a hack that add some smart guessing to accept them also. More documentation is also added about the problems. The long term fix should be properly parsing the device tree, as it is done in the kernel already.

axel-h commented:
- Merge things from https://github.com/seL4/camkes-tool/pull/28 here and pass the architecture as parameter
- split the commit up
- write a changelog
- PR request https://github.com/seL4/global-components/pull/23 must be merged with this as the function signatures change

kent-mcleod commented on camkes/templates/macros.py:
```diff
@@ -686,32 +686,105 @@ def virtqueue_get_client_id(composition, end, configuration):
 
 
-def parse_dtb_node_interrupts(node, max_num_interrupts):
+def parse_dtb_node_interrupts(node, max_num_interrupts, arch):
```
What would it look like if you split this into two sub-functions, one for arm and one for riscv. Given that they have different deserialization formats, would this be simpler?

axel-h replied:
I am not sure if this is really helpful in the current state. My understanding is, that the interrupt description format is architecture agnostic. For RISC-V, there seems no use case for the 3-value format, because the 'type' has nothing specified for RISC-V. The 2-value format could be used though if one wants to describe the trigger. All we have at the moment are the SiFive device trees, that is a too small base to judge on what to do. Our MiG-V device tree will basically follow what SiFive does and keep it to the bare minimum that seL4 needs.
I would suggest to use the current way for the near future, as this provides good flexibility and can be adapted easily to whatever device tree will come.

kent-mcleod replied:
Currently I can't easily tell that this doesn't introduce a regression for Arm just by looking at the changes.

> My understanding is, that the interrupt description format is architecture agnostic.

The description format is defined by the parent interrupt controller.  If the parent is a gicv2 then it's one format but if it's a plic0 then it's a different format.  When parsing the node to calculate which interrupts to create, the parser needs to know how to decode the specific interrupt controller type.  Unfortunately, the current parse_dtb_node_interrupts only supports gicv2/gicv3.  If you are adding support for the plic0 then it can be helpful to split the parsing logic into a plic part and gic part.

axel-h replied:
I left the ARM code in now with a comment that it is used to keep the old behavior.

kent-mcleod reviewed (commented):
The camkes-tool already gets given an import path to the capDL python library, so it shouldn't be too hard to give it access to the kernel hardware_gen python lib that would allow proper irq parsing.

axel-h commented on camkes/templates/macros.py:
```diff
@@ -686,32 +686,105 @@ def virtqueue_get_client_id(composition, end, configuration):
+            return []
+
+    # The proper way to figure out how may value per interrupt are in used is
+    # checking the '#interrupt-cells#' property of the interrupt controller.
```
```suggestion
    # checking the '#interrupt-cells' property of the interrupt controller.
```

axel-h commented:
The test fail because https://github.com/seL4/global-components/pull/23 is needed also.

nomadeel commented:
So far, this looks ok to me. Going forward though, I agree with @kent-mcleod  that we should probably take a look re-using the devicetree parsing modules inside the kernel for parsing bits and pieces of the devicetree for CAmkES. The parsing functions here could also use a reference to the entire devicetree as well as we need to be able to look through the entire tree for figuring out what the interrupt controller is for a particular device.

axel-h commented:
- Rebased to master, fixed typos in comments. 
- Is there an agreement to merge this for now? It provides a bit more flexibility at least and unblocks some issue on RISC-V platforms that are much more difficult to handle otherwise.

nomadeel commented:
Yep, I'll merge this for you.
